WIP - use timeslice reduction counting

potatosalad opened the pull request:
This pull request isn't ready to be merged yet, but I wanted to submit it to get some feedback on whether you find these changes useful or not.  I'll try to edit this pull request sometime tomorrow with a better write up, but basically&hellip;

I split each of your CRC functions into 3 stages:

1. init
2. update
3. final

So, there are `:crc_nif.crc_8_init/1`, `:crc_nif.crc_8_update/2`, and `:crc_nif.crc_8_final/1` functions as well as a `:crc_nif.crc_8/2` function.  I left your algorithms mostly untouched, but rearranged the code a little (hopefully that's okay).

All of the tests pass and you can feed it fairly large binaries without blocking any schedulers:

```elixir
input = :binary.copy("0123456789", 1024 * 1024 * 10) # roughly 100MB

# Blocking version before pull request:
{:reductions, r0} = :erlang.process_info(self(), :reductions)
CRC.crc_8(input)
{:reductions, r1} = :erlang.process_info(self(), :reductions)
reductions = r1 - r0
# => ~200

# Non-blocking version after pull request:
{:reductions, r0} = :erlang.process_info(self(), :reductions)
CRC.crc_8(input)
{:reductions, r1} = :erlang.process_info(self(), :reductions)
reductions = r1 - r0
# => ~750000
```

TattdCodeMonkey commented:
This is amazing, I'll try to find time to dig into your changes this week. 

This has been on my TODO list for awhile and I've had a hard time finding the time to dig into it.

potatosalad commented:
I neglected to mention one of the tracing features in the code to help explain how maby bytes are processed per call.  Also, odd note, github doesn't seem to be updating this pull request with my latest commits as of this morning.

If you enable tracing by uncommenting [this line in `xnif_slice.h`](https://github.com/potatosalad/crc/blob/74cfcbfc61a02c8373d426738d4bd66aa8b78ac4/c_src/nif/xnif_slice.h#L116) (you may also need to run `mix clean`), it will print out some of the internal details of the reductions used per slice:

```elixir
iex> input = :binary.copy("0123456789", 1024 * 200) # ~2MB
iex> CRC.crc_8(input)
199
xnif_slice_trap:xnif_slice.c:137 phase = 1, offset = 0, reductions = 20000
xnif_slice_trap:xnif_slice.c:137 phase = 1, offset = 20000, reductions = 20000
xnif_slice_trap:xnif_slice.c:137 phase = 1, offset = 40000, reductions = 20000
xnif_slice_trap:xnif_slice.c:137 phase = 1, offset = 60000, reductions = 20000
xnif_slice_trap:xnif_slice.c:173 trap->total = 110
xnif_slice_trap:xnif_slice.c:137 phase = 1, offset = 80000, reductions = 72000
xnif_slice_trap:xnif_slice.c:137 phase = 1, offset = 152000, reductions = 72000
xnif_slice_trap:xnif_slice.c:137 phase = 1, offset = 224000, reductions = 72000
xnif_slice_trap:xnif_slice.c:137 phase = 1, offset = 296000, reductions = 72000
xnif_slice_trap:xnif_slice.c:173 trap->total = 127
xnif_slice_trap:xnif_slice.c:137 phase = 1, offset = 368000, reductions = 210240
xnif_slice_trap:xnif_slice.c:137 phase = 1, offset = 578240, reductions = 210240
xnif_slice_trap:xnif_slice.c:173 trap->total = 139
xnif_slice_trap:xnif_slice.c:137 phase = 1, offset = 788480, reductions = 256493
xnif_slice_trap:xnif_slice.c:173 trap->total = 104
xnif_slice_trap:xnif_slice.c:137 phase = 1, offset = 1044973, reductions = 246234
xnif_slice_trap:xnif_slice.c:137 phase = 1, offset = 1291207, reductions = 246234
xnif_slice_trap:xnif_slice.c:173 trap->total = 142
xnif_slice_trap:xnif_slice.c:137 phase = 1, offset = 1537441, reductions = 285632
xnif_slice_trap:xnif_slice.c:137 phase = 1, offset = 1823073, reductions = 224927
```

By default, the max number of bytes processed per slice is ~20KB.

So for the first 4 lines of trace output you can see that it was able to process ~80KB worth of bytes in the time it took 1 millisecond of time to pass.

Since it went over by ~10% of that 1 millisecond, however, so it adjusted the max per slice to ~72KB, which is what we see on lines 6-9.

Everything from that point forward are the different recalculations that happen based on current system load and adjusted based on how much of 1 millisecond is used while processing.

I didn't come up with this strategy, but it's based on the work done by Steve Vinoski in [vinoski/bitwise](https://github.com/vinoski/bitwise).  Steve is actually one of the reasons we have dirty NIFs and timeslice funcitonality built into the NIF API.

There's probably some improvements that could be made (for example, even 1 millisecond might be too long a NIF to block), but I wanted to mention all of that in case you were curious about how it "decides on max binary size" as mentioned in #7.

potatosalad commented:
Okay, I think I'm done fiddling with things for a while.

I added a timeout adjustment for the timeslice calculations [here in `xnif_slice.h`](https://github.com/potatosalad/crc/blob/ab3061de3aba3889731a987a4ca0bd615d27c38c/c_src/nif/xnif_slice.h#L32).  By default, it's set to 500µs or 0.5ms.

I also added a [`XNIF_SLICE_MIN_PER_SLICE`](https://github.com/potatosalad/crc/blob/ab3061de3aba3889731a987a4ca0bd615d27c38c/c_src/nif/xnif_slice.h#L29) setting to prevent the adjustment algorithm from bottoming out at 0 (and then looping indefinitely with 0 reductions allowed).

I also switched from using [`gettimeofday(2)`](http://man7.org/linux/man-pages/man2/gettimeofday.2.html) to [`enif_monotonic_time`](http://erlang.org/doc/man/erl_nif.html#enif_monotonic_time) for a little more control over how the clock difference is calculated.

fire commented:
Can you support crc32c castagnoli?

TattdCodeMonkey commented:
>I'm not sure whether some of the CRC implementations are "correct"

Yea... I tend to try and validate against the two JS implementations I have linked in the [README & docs](https://github.com/TattdCodeMonkey/crc/blob/master/lib/crc.ex#L12-L13) but its still only as correct as they are. With a recent PR Kermit they conflicted but it looked like an Endian conversion issue. 

I started this because I needed the CRC-16 CCITT for a personal project and slowly added other algorithms as I had time or people PR'd them. But verification is definitely tricky. CRC being most prevalent in hardware & embedded space, and then that industry seems to have low OSS adoption.

TattdCodeMonkey commented:
> Possible Improvements
>
> The complexity of the NIF functions could potentially be reduced by generalizing the parameters passed for each CRC operation. Using magic references which are bound to the actual parameters for the CRC function would also mirror the way crypto handles hash and HMAC operations.

That is very appealing and has me second guessing my v1 release plan. I had been planning on moving to v1.0 once I implemented time slicing (which I don't have to do now 🎉). But I might chew on this suggestion a bit because it would be a much nicer interface. And then I could have some hot paths with tables compiled for the more common algorithms 🤔. I'm going to think on this some before pushing any buttons and moving to v1.

TattdCodeMonkey commented:
The move to delegating everything to the erlang module gave me a little bit of heartburn just b/c I've written almost no erlang, but in the end I think it makes the library more robust and easy to use from both elixir and erlang. 

I can mostly follow the Makefile so I'll assume it's better. I had just copy pasted the previous one from a mix of projects just to get it to work, so it being written by someone with any experience with makefiles is most likely a significant improvement.

@potatosalad Overall this is amazing from my PoV, I'm going to try and do some more testing tonight and tomorrow. But barring anything major I'll probably merge this later this week. And get v0.7 out on hex.

potatosalad commented:
> The move to delegating everything to the erlang module gave me a little bit of heartburn just b/c I've written almost no erlang, but in the end I think it makes the library more robust and easy to use from both elixir and erlang.

I understand, and it really is up to you on that front.  I do have an Erlang-only project that would benefit from being able to import the same crc package from hex that my other Elixir projects can, but I understand the apprehension.  If we were to simplify the interface, however, you'll probably never touch the actual NIF call path on the Elixir or Erlang side of things.  There'd just be `crc/2`, `crc_init/1`, `crc_update/2`, and `crc_final/1` (or something similar to that).

> But I might chew on this suggestion a bit because it would be a much nicer interface. And then I could have some hot paths with tables compiled for the more common algorithms :thinking:. I'm going to think on this some before pushing any buttons and moving to v1.

I'd be happy to submit a separate pull request of what that might look like that we could play around with until it looks right.

TattdCodeMonkey commented:
@potatosalad I'm good with whats here. I'm going to merge and and prep a v0.7 release. if you want to start a  `crc/2`, `crc_init/1`, `crc_update/2`, and `crc_final/1` PR we can work through that.

I'll add a specific issue on the repo, and if your interesting I can just add you as a contributor and you can work on a branch in the repo.

TattdCodeMonkey commented:
Closes #4 #5 #6 & #7

potatosalad commented:
> if your interesting I can just add you as a contributor and you can work on a branch in the repo.

That could work, it would make it easier to adjust things from either side.  I have a partial solution working locally I was planning on pushing up soon.
